### PR TITLE
Allow GET/SET custom config param in Z-Wave device configuration

### PR DESCRIPTION
**Pull request by MindFreeze**
## Proposed change

UI form for custom configuration parameters
Requires https://github.com/home-assistant-libs/zwave-js-server-python/pull/1093
and https://github.com/home-assistant-libs/zwave-js-server-python/pull/1080

Core PR: https://github.com/home-assistant/core/pull/129332

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
+import { LitElement, html, css, CSSResultGroup, nothing } from "lit";
+import { customElement, property, state } from "lit/decorators";
+import "@material/mwc-button/mwc-button";
+import { mdiCloseCircle } from "@mdi/js";
+import "../../../../../components/ha-textfield";
+import "../../../../../components/ha-select";
+import { HomeAssistant } from "../../../../../types";
```
Comment (wendevlin) [inline src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:7]:
```suggestion
import type { HomeAssistant } from "../../../../../types";
```

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
+import { LitElement, html, css, CSSResultGroup, nothing } from "lit";
```
Comment (wendevlin) [inline src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:1]:
```suggestion
import { type LitElement, html, css, type CSSResultGroup, nothing } from "lit";
```

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+          @input=${this._customParamNumberChanged}
+        ></ha-textfield>
+        <ha-select
+          .label=${this.hass.localize(
+            "ui.panel.config.zwave_js.node_config.size"
+          )}
+          .value=${this._valueSize}
+          @selected=${this._customValueSizeChanged}
+        >
+          <ha-list-item value="0"></ha-list-item>
```
Comment (wendevlin) [inline src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:42]: Is here the label missing or should it be empty when 0?

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+          @selected=${this._customValueFormatChanged}
+        >
+          <ha-list-item value="0">Signed</ha-list-item>
+          <ha-list-item value="1">Unsigned</ha-list-item>
+          <ha-list-item value="2">Enumerated</ha-list-item>
+          <ha-list-item value="3">Bitfield</ha-list-item>
+        </ha-select>
+      </div>
+      <div class="custom-config-buttons">
+        <mwc-button @click=${this._getCustomConfigValue}>
```
Comment (wendevlin) [inline src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:68]: please us `ha-button` instead

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts [hunk 5]
```diff
@@ -0,0 +1,182 @@
+import { LitElement, html, css, CSSResultGroup, nothing } from "lit";
+import { customElement, property, state } from "lit/decorators";
+import "@material/mwc-button/mwc-button";
+import { mdiCloseCircle } from "@mdi/js";
+import "../../../../../components/ha-textfield";
+import "../../../../../components/ha-select";
+import { HomeAssistant } from "../../../../../types";
+
+@customElement("zwave_js-custom-param")
```
Comment (wendevlin) [inline src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:9]:
```suggestion
@customElement("zwave-js-custom-param")
```

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      .error-icon {
+        margin-right: 8px;
+      }
+    `;
+  }
+}
+
+declare global {
+  interface HTMLElementTagNameMap {
+    "zwave_js-custom-param": ZWaveJSCustomParam;
```
Comment (wendevlin) [inline src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:180]:
```suggestion
    "zwave-js-custom-param": ZWaveJSCustomParam;
```

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-node-config.ts [hunk 7]
```diff
@@ -179,6 +180,17 @@ class ZWaveJSNodeConfig extends LitElement {
...
+          <h3>
+            ${this.hass.localize(
+              "ui.panel.config.zwave_js.node_config.custom_config"
+            )}
+          </h3>
+          <ha-card class="custom-config">
+            <zwave_js-custom-param
+              .hass=${this.hass}
+              .deviceId=${this.deviceId}
+            ></zwave_js-custom-param>
```
Comment (wendevlin) [inline src/panels/config/integrations/integration-panels/zwave_js/zwave_js-node-config.ts:192]:
```suggestion
            <zwave-js-custom-param
              .hass=${this.hass}
              .deviceId=${this.deviceId}
            ></zwave-js-custom-param>
```

**Review by wendevlin — CHANGES_REQUESTED** [review]

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts (hunk 3 above)
Comment (MindFreeze) [reply src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:42]: it should be empty

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts (hunk 5 above)
Comment (MindFreeze) [reply src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:9]: The underscore is used in all the other zwave_js components

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+        .custom-config-form {
+          flex-wrap: nowrap;
+        }
+
+        ha-textfield,
+        ha-select {
+          flex-basis: 0;
+        }
+      }
+
```
Comment (wendevlin) [inline src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:203]:
```suggestion
.custom-config-buttons {
  display: flex;
  justify-content: flex-end;
}
```

As far as I understood our design guidelines action buttons should be placed on the bottom right.

File: src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+          .label=${this.hass.localize(
+            "ui.panel.config.zwave_js.node_config.format"
+          )}
+          .value=${String(this._valueFormat)}
+          @selected=${this._customValueFormatChanged}
+        >
+          <ha-list-item value="0">Signed</ha-list-item>
+          <ha-list-item value="1">Unsigned</ha-list-item>
+          <ha-list-item value="2">Enumerated</ha-list-item>
+          <ha-list-item value="3">Bitfield</ha-list-item>
```
Comment (wendevlin) [inline src/panels/config/integrations/integration-panels/zwave_js/zwave_js-custom-param.ts:69]: Please add translations

**Review by wendevlin — CHANGES_REQUESTED** [review]
I am not able to test it with core, because I need z-wave server 39.

- What about using `ha-progress-button` or add an additional success message?
- Could you maybe add some kind of intro/description what the `Custom configuration` is and what I can use it for? Or a reference to a documenation?

**Review by wendevlin — DISMISSED** [review]

**Review by wendevlin — DISMISSED** [review]
I wasn't able to set/get any property, but I don't know what parameters to set :laughing: 

LGTM :+1:

**Review by wendevlin — APPROVED** [review]

**Pull request merged**